### PR TITLE
Parse dot-notation validation rules for nested array shape inference

### LINKED ISSUE #624: Parse dot-notation validation rules for nested array shape inference
### Problem:
When using Laravel's dot notation for nested arrays or wildcards (e.g., author.name or comments.*.id), Psalm loses the context of the deeply nested structure. It typically falls back to a generic array<string, mixed> or fails to map the nested shapes correctly.

```
public function store(Request $request)
{
    $data = $request->validate([
        'title' => 'required|string',
        'author.name' => 'required|string',
        'author.email' => 'required|email',
        'tags.*.id' => 'required|integer',
    ]);

    // Psalm currently infers something like: array<string, mixed> if author and tags are specified explicitly as 'array'
    // Or it flattens the keys literally instead of nesting them if not.
    
    // We want Psalm to infer:
    // array{
    //     title: string, 
    //     author: array{name: string, email: string}, 
    //     tags: list<array{id: int}>
    // }
}
```

### Proposed solution
Introduce a parser to convert dot-notation strings into a structured tree format. By evaluating these keys as a tree rather than flat strings, the plugin can dynamically construct the correct, nested Psalm array shapes (including handling * wildcards as lists/iterables).

I've prepared a PR for this feature: #625

### PR DESCRIPTION
**Pull request by MDG11**
## What does this PR do?
PR fixes #624 
Introduces a parser to convert Laravel's dot-notation validation rules into a structured tree format. This allows the plugin to accurately resolve and infer deeply nested array shapes returned by ``$request->validate()`` and ``Validator::validated()``, bridging the gap between Laravel's validation strings and Psalm's static type system.

## Before:
Validating nested data with dot notation (e.g., ``['author.name' => 'required|string']``) resulted in Psalm falling back to generic ``array<string, mixed>`` or losing the nested type context entirely.

## After:
The plugin successfully parses the dot notation into a nested tree structure. Psalm now infers the correct array shapes, such as ``array{author: array{name: string}}``, and correctly handles wildcard array notations (e.g., ``comments.*.id``).

## How it was tested                                                                                                                                                                              
                                                                                                                                                                                                    
  Testing uses the project's `.phpt`-based type-assertion harness                                                                                                                                   
  (`tests/Type/tests/ValidationTypeNarrowingTest.phpt`), which runs Psalm over                                                                                                                      
  inline PHP snippets and asserts exact inferred types via `@psalm-check-type-exact`.                                                                                                               
                                                                                                                                                                                                    
  ### New test cases added (lines 153–253)                                                                                                                                                          
                                                                                                                                                                                                    
  | Test function | What it covers |                                                                                                                                                                
  |---|---|                                                                                                                                                                                       
  | `testNestedDotNotationShape` | Two dot-notation keys under the same prefix (`address.city`, `address.zip`) collapse into `array{address: array{city: string, zip: string}}` |                   
  | `testMixedFlatAndNestedShape` | Flat and nested keys coexist: `name` stays top-level while `address.city` / `address.street` nest under `address` |                                             
  | `testDeepNestingShape` | Three-level nesting (`billing.address.city`) produces `array{billing: array{address: array{city: string, street: string}}}` |                                          
  | `testOptionalNestedField` | `sometimes` on a nested key marks it optional (`phone?: string`) while the sibling stays required |                                                                 
  | `testNullableNestedField` | `nullable` on a nested key produces `bio?: null\|string`; the field is also marked possibly-undefined because Laravel only includes nullable fields when they are   
  present |                                                                                                                                                                                         
  | `testNestedDotNotationSingleField` | `validated('address.city')` (single-key overload) still resolves to `string` for a dot-notation key |                                                    
  | `testInlineValidateDotNotation` | Inline `$request->validate(['user.name' => ..., 'user.email' => ...])` produces the same nested shape without a `FormRequest` subclass |                      
                                                                                                                                                                                                    
  ### Implementation path exercised                                                                                                                                                                 
                                                                                                                                                                                                    
  ValidationRuleAnalyzer is unchanged — dot-notation keys (e.g. address.city) pass through resolveRules() as regular flat entries and are returned as array<string, ResolvedRule>.                  
                                                                                                                                                                                                    
  The nesting logic lives entirely in ValidatedTypeHandler::buildArrayShape():                                                                                                                      
                                                                                                                                                                                                    
  1. For each flat key, ``explode('.', $field)`` splits it into segments (``['address', 'city']``).                                                                                                         
  2. Segments are passed to the recursive ``insertIntoTree()``, which shifts the head segment as the current tree key and recurses into ``$tree[$key]`` until the last segment, where it stores the       
  ``ResolvedRule`` as a leaf.                                                                                                                                                                           
  3. Once the full tree is built, ``buildUnionFromTree()`` walks it depth-first: leaf nodes become typed Union properties (with ``setPossiblyUndefined(true)`` for optional fields), nested nodes recurse 
  and are wrapped in a ``TKeyedArray``, producing the final nested Psalm shape.
                                                                                                                         
## Checklist
- [X] Tests cover the change (type test in `tests/Type/` and/or unit test in `tests/Unit/`)

### CODE REVIEW

**Review by alies-dev — COMMENTED** [review]
Great work on this PR, @MDG11 ! The dot-notation parsing approach is solid — building an intermediate tree from `explode('.')` segments and then recursively converting it to nested `TKeyedArray` shapes is the right design. The test coverage is thorough (7 new cases covering flat+nested, deep nesting, optional, nullable, single-field access, and inline validate).

A few things I noticed:

### Bug: missing `Union` wrapper (inline comment below)

`Type::getListAtomic()` returns `TKeyedArray|TArray` (an atomic), not a `Union`. The `if` branch wraps it correctly, but the `else` fallback doesn't. See inline comment on line 388.

### Style: global function backslash prefix

The project convention requires backslash-prefixed global function calls (`\explode()`, `\array_shift()`, `\is_array()`). Running `composer cs` should auto-fix all of these.

### Simplification: wildcard block in `insertIntoTree` (lines 330-338)

The `$key === '*'` branch does the exact same thing as the generic path below it (ensure array exists, recurse). Since `ValidationRuleAnalyzer::resolveRules()` already pre-resolves all wildcard patterns before rules reach `buildArrayShape()`, this entire block is unreachable and can be safely removed. The outer condition `$segments[0] === '*'` also has no effect — when it's true but `$key !== '*'`, execution falls through to the generic case anyway.

### Nice to have: test ideas for future PRs

The core cases are well covered. A few edge cases that could be interesting follow-ups (no need to block this PR):

- `safe(['address.city'])` — verifies partial shape extraction works with dot-notation keys
- `items.*.address.city` — wildcard children combined with dot-notation nesting (the wildcard pre-processing in `resolveRules()` produces flat keys like `address.city` inside the wildcard children map, so this would produce `list<array{address.city: string}>` rather than the expected nested shape — worth a test to document this)
- Parent + child rules (`'address' => 'required|array'` + `'address.city' => 'required|string'`) — common Laravel pattern worth documenting behavior for

Thanks for the contribution!

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 1]
```diff
@@ -289,29 +289,119 @@ private static function resolveFieldType(
...
+
+            if (isset($value['*']) && is_array($value['*'])) {
+                $itemShape = self::buildUnionFromTree($value['*']);
+
+                if ($itemShape !== null) {
+                    $properties[$key] = new Union([
+                        Type::getListAtomic($itemShape),
+                    ]);
+                } else {
+                    $properties[$key] = Type::getListAtomic(Type::getMixed());
```
Comment (alies-dev) [inline src/Handlers/Validation/ValidatedTypeHandler.php]: Bug: `Type::getListAtomic()` returns `TKeyedArray|TArray` (an atomic type), but `$properties[$key]` expects a `Union`. The `if` branch on line 384 correctly wraps it in `new Union([...])` — this `else` branch should do the same.

```suggestion
                    $properties[$key] = new Union([Type::getListAtomic(Type::getMixed())]);
```

File: tests/Type/tests/ValidationTypeNarrowingTest.phpt [hunk 2]
```diff
@@ -149,5 +149,107 @@ function testInlineValidateArrayFormat(\Illuminate\Http\Request $request): void
     ]);
     /** @psalm-check-type-exact $_data = array{active?: '0'|'1'|0|1|bool, id: string} */
 }
+
+class NestedAddressRequest extends FormRequest
+{
+    public function rules(): array
+    {
+        return [
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply tests/Type/tests/ValidationTypeNarrowingTest.phpt:157]: Good catch. In the follow-up commit (c1761b4) we restructured the node as `ValidationRuleNode` with a separate `$rule` property, so a parent key having both its own rule and nested children is now properly modeled (the `$rule` is preserved alongside `$children`).

The test gaps you mention are real — we have documented them as follow-up items. Specifically: parent-with-own-rule (e.g., `'user' => 'required |array'` alongside `'user.name'`) and all-optional-descendants affecting parent requiredness. Will add those in subsequent commits on this branch.

File: src/Handlers/Validation/ValidatedTypeHandler.php [hunk 3]
```diff
@@ -289,29 +289,111 @@ private static function resolveFieldType(
...
+                        Type::getListAtomic(Type::getMixed())
+                    ]);
+                }
+                continue;
+            }
+
+            $nested = self::buildUnionFromTree($value);
+
+            if ($nested !== null) {
+                $properties[$key] = $nested;
```
[the remark this answers is not in the document]
Comment (alies-dev) [reply src/Handlers/Validation/ValidatedTypeHandler.php:388]: Acknowledged — this is now documented as a known limitation in the `buildUnionFromTree()` docblock (see commit c1761b4, lines 338-341).

The suggested fix of unconditionally marking all intermediate nodes as `possibly_undefined` would overcorrect: when any descendant is `required`, Laravel guarantees the parent key is present. The correct fix would be to recursively check whether ANY descendant is required, which adds complexity for an edge case. Deferred for now, tracked for follow-up.
